resolve.go

```diff
@@ -41,34 +44,39 @@ func (h *handler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
 }
 
 func (h *handler) resolveDnsNames(names []string) (*dns.Msg, error) {
-	h.log.Info("forwarding request", "resolver", dnsServer)
+	h.log.Info("forwarding request", "resolver", inputConfig.dnsServer)
 
 	dnsClient := h.dnsClient
 
 	for _, name := range names {
-		retried := false
+		attempt := 1
 
 	Redo:
 		dnsRequest := new(dns.Msg)
 		dnsRequest.SetQuestion(name, dns.TypeA)
 		dnsRequest.SetEdns0(4096, true)
 
 		h.log.Info("doing dns lookup", "req", dnsRequest)
-		ans, rtt, err := dnsClient.Exchange(dnsRequest, dnsServer)
+		ans, rtt, err := dnsClient.Exchange(dnsRequest, inputConfig.dnsServer)
 		if err != nil {
-			if !retried {
-				// Add backoff
-				retried = true
+			if inputConfig.MaxRetries >= attempt {
+				backOffTime := h.backoff.Next(attempt)
+				time.Sleep(backOffTime)
+				h.log.Info(fmt.Sprintf("Received Error: %v. Retrying dns lookup. Attempt: %v after %v", err, attempt, backOffTime))
+				attempt++
 				goto Redo
 			}
 			h.log.Error(err, "cannot resolve dns", "dns-request", dnsRequest)
 			return nil, err
 		}
 		if ans.MsgHdr.Truncated {
-			if !retried {
-				retried = true
+			if inputConfig.MaxRetries >= attempt {
```

Comment:
Truncated responses shouldn't be considered as error, so shouldn't go through the backoff logic, we should immediately reissue the request via TCP

Reply:
Got it Suresh. Will change.

Reply:
Done